Game/Plugins/MetricsServiceProvider/Source/MetricsServiceProvider/Private/MetricsServiceModule.cpp

```diff
@@ -0,0 +1,640 @@
+{
+#if WITH_EDITOR
+	return "editor";
+#else
+	if (GWorld)
+	{
+		switch (GWorld->GetNetMode())
+		{
+			case NM_Client:
+				return "client";
+			case NM_Standalone:
+				return "standalone";
+			default:
+				return "server";
+		}
+	}
+	else
+	{
+		// If we don't have a world, we are neither client or server
+		// so standalone makes the most sense(and can be the case on client shutdown)
+		return "standalone";
+	}
+#endif
+}
+
+void FAnalyticsMetricsServiceModule::StartupModule()
+{
+}
+
+void FAnalyticsMetricsServiceModule::ShutdownModule()
+{
+	FAnalyticsProviderMetrics::Destroy();
+}
+
+TSharedPtr<IAnalyticsProvider> FAnalyticsMetricsServiceModule::CreateAnalyticsProvider(const FAnalyticsProviderConfigurationDelegate& GetConfigValue) const
+{
+	// Use our project name for our analytics endpoints.  By default UE4 uses the compiled build type of the game.
+	FString ProjectName;
+	FString AnalyticsSection;
+	if (FParse::Value(FCommandLine::Get(), TEXT("projectName"), ProjectName))
+	{
+		AnalyticsSection = TEXT("Analytics_") + ProjectName;
+	}
+	else
+	{
+		AnalyticsSection = TEXT("Analytics");
+	}
+
+	const FString Key = FAnalytics::Get().GetConfigValueFromIni(GEngineIni, AnalyticsSection, TEXT("ApiKey"), true);
+	const FString ApiEndpoint = FAnalytics::Get().GetConfigValueFromIni(GEngineIni, AnalyticsSection, TEXT("ApiEndpoint"), true);
+	const int32 BatchSize = FCString::Atoi(*FAnalytics::Get().GetConfigValueFromIni(GEngineIni, AnalyticsSection, TEXT("BatchSize"), true));
+	const FString BinaryKey = FAnalytics::Get().GetConfigValueFromIni(GEngineIni, AnalyticsSection, TEXT("BinaryApiKey"), true);
+	const FString BinaryEndpoint = FAnalytics::Get().GetConfigValueFromIni(GEngineIni, AnalyticsSection, TEXT("BinaryEndpoint"), true);
+	const float MaxAge = FCString::Atof(*FAnalytics::Get().GetConfigValueFromIni(GEngineIni, AnalyticsSection, TEXT("MaxAge"), true));
+
+	return FAnalyticsProviderMetrics::Create(Key, ApiEndpoint, BinaryKey, BinaryEndpoint, BatchSize, MaxAge);
+}
+
+// Provider
+const FString FAnalyticsProviderMetrics::InvalidUserId = TEXT("User Not Set");
+
+// The maximum number of retries we will attempt.
+const uint32 RetryLimitCount = 10;
+// The maximum number of seconds we will try to retry inside of.
+const uint32 RetryLimitTime = 60;
+
+FAnalyticsProviderMetrics::FAnalyticsProviderMetrics(const FString& Key, const FString& ApiEndpoint, const FString& BinaryKey, const FString& BinaryEndpoint, const int32 BatchSize, const float MaxAge) :
+	ApiKey(Key),
+	BinaryApiKey(BinaryKey),
+	EndPointURL(ApiEndpoint),
+	BinaryEndpointURL(BinaryEndpoint),
+	BatchSizeThreshold(BatchSize),
+	EngineVersion(*FEngineVersion::Current().ToString()),
+	EventEnvironment(LexToString(FApp::GetBuildConfiguration())),
+	MaxAgeThresholdSeconds(MaxAge)
+{
+	HttpRetryManager = MakeShared<FHttpRetrySystem::FManager>(
+		FHttpRetrySystem::FRetryLimitCountSetting(),
+		FHttpRetrySystem::FRetryTimeoutRelativeSecondsSetting());
+
+	Http = &FHttpModule::Get();
```

Comment:
Seems we only check if it's NULL in flush events

Reply:
To check if we have access for this module

Reply:
Hmm if Http is null in the FlushEvents we just try again to get it again. It seems like we could continually fail to get access to Http and it would not have any affect on this code 🤔 If we can test this code with this code removed that might be good. Seems odd to me. Maybe there is a good reason it's there that I am to dumb to see...